Fix encoding bug in `gensim.scripts.word2vec2tensor`. Fix #1958

AakaashRao opened the pull request:
This PR fixes #1958, which noted an issue with the Python 3 (but not Python 2) version of the `word2vec2tensor` function. The problematic line was attempting to encode a string in `utf-8` and write the resulting bytestring to file, causing an error; this revised code simply writes the original string. I have confirmed that the code runs as expected on Python 2.7 and Python 3.6.

CC @menshikh-iv.

menshikh-iv commented:
Hello @AakaashRao 

1. It's better to write all as bytes always (this is some our convention)
2. `.DS_store` is unrelated file, please remove it
3. Definitely, need to add a tests for this functionality 
4. Replace `open` to `smart_open`

menshikh-iv commented on gensim/scripts/word2vec2tensor.py:
```diff
@@ -63,22 +64,25 @@ def word2vec2tensor(word2vec_model_path, tensor_filename, binary=False):
         True if input file in binary format.
 
     """
+
```
style-nitpick: no need an empty line here

menshikh-iv commented on gensim/scripts/word2vec2tensor.py:
```diff
@@ -63,22 +64,25 @@ def word2vec2tensor(word2vec_model_path, tensor_filename, binary=False):
+        with utils.smart_open(outfiletsvmeta, 'wb') as file_metadata:
             for word in model.index2word:
-                file_metadata.write(gensim.utils.to_utf8(word) + gensim.utils.to_utf8('\n'))
+                write = gensim.utils.any2utf8(word) + gensim.utils.any2utf8('\n')
```
no needed changes (this is almost same), you can return an old variant (except for last line in cycle)

menshikh-iv commented on gensim/test/test_word2vec2tensor.py:
```diff
@@ -0,0 +1,79 @@
+
+class TestWord2Vec2Tensor(unittest.TestCase):
+    def setUp(self):
+        # self.datapath = datapath('test_word2vec.txt')
```
this is right way how to retrieve path

menshikh-iv commented on gensim/test/test_word2vec2tensor.py:
```diff
@@ -0,0 +1,79 @@
+
+    def testConversion(self):
+        check_output(args=[
+            sys.executable, '-m', 'scripts.word2vec2tensor', '--input',
```
you should test changed function `word2vec2tensor`, not all script

menshikh-iv commented on gensim/test/test_word2vec2tensor.py:
```diff
@@ -0,0 +1,79 @@
+                )
+        try:
+            with smart_open(self.tensor_file, 'rb') as f:
+                self.vectors = f.readlines()
```
I don't catch, why you need `self` here (same question for `number_words`, `metadata`, etc)?

menshikh-iv commented on gensim/test/test_word2vec2tensor.py:
```diff
@@ -0,0 +1,79 @@
+
+        # test that the converted model loads successfully
+        self.test_model = gensim.models.KeyedVectors.load_word2vec_format(self.vector_file, binary=False)
+        self.assertTrue(numpy.allclose(self.test_model.n_similarity(['the', 'and'], ['and', 'the']), 1.0))
```
+1 for this test, but -1 for this type of assertion.
If you want to check that models are equivalent - you should be sure, that vectors for each word are equivalent.

menshikh-iv commented on gensim/test/test_word2vec2tensor.py:
```diff
@@ -0,0 +1,79 @@
+
+
+if __name__ == '__main__':
+    logging.basicConfig(format='%(asctime)s : %(levelname)s : %(message)s', level=logging.DEBUG)
```
please use `logging.INFO` instead of `DEBUG`

AakaashRao commented:
Thanks @menshikh-iv for the feedback. I've addressed your points and written some tests. A couple of quick points/questions about the tests:

1. I've added some test word2vec vectors into `test_data` as `test_data/test_word2vec.txt` because I couldn't find any vectors saved already. Let me know if this is incorrect.
2. Even in a `virtualenv`, I could not get the `datapath` function -- which is supposed to prepend the `test_data` directory to a filename -- to look in the `test_data` folder of my development version of gensim (containing the new `test_word2vec.txt` file) rather than in the `test_data` in my installed gensim. For now I have replaced it with a temporary fix, but I would appreciate any guidance -- I'm sure I'm missing something obvious.

menshikh-iv commented:
@AakaashRao 
1. For example https://github.com/RaRe-Technologies/gensim/blob/develop/gensim/test/test_data/word2vec_pre_kv_c (large) or https://github.com/RaRe-Technologies/gensim/blob/develop/gensim/test/test_data/high_precision.kv.txt (small)
2. Yes, you have a very old fork, for fixing this, you should to fetch fresh upstream and merge `upstream/develop` to your current branch.

you stay in PR branch
```bash
git remote add upstream https://github.com/RaRe-Technologies/gensim.git
git fetch upstream
git merge upstream/develop
```

AakaashRao commented:
Thanks @menshikh-iv. I've addressed your review. Like before, I had some problems getting Python to import the current working version of `word2vec2tensor.py` instead of the version I had installed from the master version of Gensim (even when using the virtualenv). As a temporary fix, I placed a `scripts` folder containing `word2vec2tensor.py` in `test`. I would really appreciate if you could tell me how to fix this issue.

menshikh-iv commented on gensim/scripts/word2vec2tensor.py:
```diff
@@ -67,12 +68,12 @@ def word2vec2tensor(word2vec_model_path, tensor_filename, binary=False):
+        with utils.smart_open(outfiletsvmeta, 'wb') as file_metadata:
             for word in model.index2word:
-                file_metadata.write(gensim.utils.to_utf8(word) + gensim.utils.to_utf8('\n'))
+                file_metadata.write(gensim.utils.any2utf8(word) + gensim.utils.any2utf8('\n'))
```
why `any2utf8` instead of `to_utf8`?

menshikh-iv commented on gensim/test/test_scripts.py:
```diff
@@ -103,6 +107,75 @@ def test_segment_and_write_all_articles(self):
+                vectors = f.readlines()
+        except FileNotFoundError:
+            self.fail(
+                'Tensor file %s creation failed. \
```
carriage-return  (here and everywhere)

menshikh-iv commented on gensim/test/scripts/word2vec2tensor.py:
```diff
@@ -0,0 +1,95 @@
+#!/usr/bin/env python
```
`from gensim.scripts.word2vec2tensor import word2vec2tensor` should work fine (without this file).

menshikh-iv commented:
Looks abandoned, I close this.
@AakaashRao feel free to re-open when you be ready to finish it.

martinnormark commented:
Oh darn, this would've been useful to get merged.

Is it only the last three review comments, that prevents this from getting merged?

FWIW, I manually fixed this in a separate conda environment by adding `decode` calls:
```python
file_metadata.write(gensim.utils.to_utf8(word).decode("utf-8") + gensim.utils.to_utf8('\n').decode("utf-8"))
```

menshikh-iv commented:
@martinnormark mostly yes, feel free to submit new PR :+1:
